### PR TITLE
Refactor CUDA copy and general copy dispatch

### PR DESCRIPTION
**Pull request by colesbury**
Copy.cu goes from 308 to 190 lines of code. In general it uses, the same
copy strategy, using cudaMempcyAsync, a pointwise kernel, or a copy
using temporary buffers. The pointwise kernel has slightly improved
performance when broadcasting due to faster index calculation.

This deletes "`s_copy_`", "`_s_copy_from`", and "`_copy_same_type_`". The only
entry-point now is "`copy_`".

A mini-benchmark is here:
https://gist.github.com/colesbury/706de1d4e8260afe046020988410b992

Before:
https://gist.github.com/colesbury/ab454b6fe3791bff420d7bcf8c041f18
After:
https://gist.github.com/colesbury/9024d242b56ab09a9ec985fa6d1620bc

Results were measured on 2.2 GHz Broadwell; no-turbo; one thread;
compiled with GCC 7.3.0. (Results are slower than typical usage due to
turbo being off.)

The only significant differences is in the CUDA [1024] -> [1024, 1024]
broadcasting copy which is ~25% faster. I don't expect a noticeable
difference in real programs.

CPU copy overhead is a tiny bit (~200 ns) faster, but I don't expect
anyone to notice that.

### CODE REVIEW

File: aten/src/ATen/native/Copy.cpp [hunk 1]
```diff
@@ -1,121 +1,126 @@
...
+
+bool copy_transpose_valid(const Tensor& self, const Tensor& src) {
   const int MIN_SZ = 60 * 60;
   return self.is_contiguous() && src.numel() != 0 && src.dim() == 2 &&
       src.stride(0) == 1 && src.stride(1) == src.size(0) &&
       self.numel() >= MIN_SZ;
 }
 
+// special case copy where tensor is contiguous and src is a transposed matrix
+// This can be generalized to most copies, but it's tricker
```
Comment (SsnL) [inline aten/src/ATen/native/Copy.cpp]: trickier :)

**Review by soumith — APPROVED** [review]
lgtm

**Comment by ezyang** [comment]
This appears to have broken XLA CI, cc @ailzhang

**Comment by colesbury** [comment]
@ezyang @ailzhang do we have an XLA contbuild?

**Comment by colesbury** [comment]
Oh, the XLA contbuild "runs" on PRs, but only checks out code (not build or tests)...

**Comment by dlibenzi** [comment]
This has broken XLA, in two ways.
One, a build issue, which we have tracked down and fixed:

https://github.com/pytorch/xla/pull/697

But there is some more trouble in there.
When we do a `variable.data.cpu()` (where `variable` is an XLA variable), we end up with this stack trace:
```
#0  at::native::to_impl (self=..., options=..., non_blocking=false) at ../aten/src/ATen/native/TensorConversions.cpp:23
#1  0x00007fffe1fbe2d6 in at::native::to (self=..., options=..., non_blocking=false, copy=false) at ../aten/src/ATen/native/TensorConversions.cpp:55
#2  0x00007fffe23dfaaf in at::TypeDefault::to (this=0x5555571353f0, self=..., options=..., non_blocking=false, copy=false) at aten/src/ATen/TypeDefault.cpp:4251
#3  0x00007fffe48fb913 in torch::autograd::VariableType::to (this=0x5555571353f0, self=..., options=..., non_blocking=false, copy=false) at ../torch/csrc/autograd/generated/VariableType_2.cpp:16201
#4  0x00007fffe902dbfa in at::Tensor::to (this=0x7fff9454c490, options=..., non_blocking=false, copy=false) at ../aten/src/ATen/core/TensorMethods.h:817
#5  0x00007fffe9020074 in torch::autograd::dispatch_to (self=..., device=..., non_blocking=false, copy=false) at ../torch/csrc/autograd/generated/python_variable_methods.cpp:300
#6  0x00007fffe8f6434b in torch::autograd::THPVariable_cpu (self=0x7fff9454c480, args=0x0) at ../torch/csrc/autograd/generated/python_variable_methods.cpp:317
```

Essentially, `variable.data` is still a Variable, which ends up calling `TypeDefault::to()` (so even if we override them, it is useless), which ends up creating a CPU tensor and asking it to `copy_()` from an XLA tensor.
Which is not going to happen.

**Comment by colesbury** [comment]
@dlibenzi I'm confused by the cpu()/to() code path you describe. The PR changes copy, but not cpu() or to(). The error in the code path you describe sounds like it happens before copy_ is called -- that to is creating a CPU tensor instead of an XLA tensor. What am I missing?

**Comment by dlibenzi** [comment]
I am not sure whether this PR changes are the culprit. Have not looked at the details.
I am guessing the removal of the `s_copy*` paths, dropped the hooks we used to make things right, WRT CPU type conversion.
We then end up with this crash:

https://pytorch.slack.com/files/UDK78U5JB/FJVMT2JSH/variable_data_cpy___crash_stack_trace.txt

Like described above, the `to_impl()` API ends up creating a CPU tensor, and asking it to `copy_()` an XLA tensor.

File: aten/src/ATen/native/Copy.cpp [hunk 2]
```diff
@@ -1,121 +1,126 @@
...
-    std::tie(b_src) = expand_inplace(self, src, "copy");
-    return s_copy_(self, b_src, non_blocking);
-  }
-  AT_ERROR("copy_() between dense and sparse Tensors is not implemented! Found self type = ",
-           self.type(), " and src type = ", src.type());
-}
 
-Tensor& _s_copy__cpu(Tensor& self, const Tensor& src, bool non_blocking) {
-  if (src.type_id() != CPUTensorId()) {
-    _s_copy_from(src, self, non_blocking);
```
Comment (dlibenzi) [inline aten/src/ATen/native/Copy.cpp]: I believe this might be the path that broke XLA.

Comment (colesbury) [reply aten/src/ATen/native/Copy.cpp]: Was `self` an XLA tensor? or did aten_xla_type hack it to work with copying into a CPU tensor?

https://github.com/pytorch/xla/blob/b7c0ab3f8cd63e490ed03ecdbef7b39ea069b8fc/torch_xla/csrc/aten_xla_bridge.cpp#L104

Comment (dlibenzi) [reply aten/src/ATen/native/Copy.cpp]: `src` was an XLA tensor

**Pull request closed without merging**